File: web/src/api/cicd.ts
import type { Pipeline, CICDJob, JobLog, Artifact, Secret, Environment, Deployment, Runner, CronSchedule } from '@/lib/types'
import { API_BASE } from '@/lib/api'

type RequestFn = <T>(endpoint: string, options?: { method?: string; body?: any }) => Promise<T>

export function createCICDApi(request: RequestFn, _baseUrl: string) {
  return {
    // Pipelines
    listPipelines: (owner: string, repo: string, page = 1, pageSize = 20) =>
      request<{ items: Pipeline[]; total: number; page: number; pageSize: number }>(
        `/repos/${owner}/${repo}/pipelines?page=${page}&page_size=${pageSize}`
      ),
    getPipeline: (owner: string, repo: string, id: number) =>
      request<Pipeline>(`/repos/${owner}/${repo}/pipelines/${id}`),
    cancelPipeline: (owner: string, repo: string, id: number) =>
      request<{ ok: boolean }>(`/repos/${owner}/${repo}/pipelines/${id}/cancel`, { method: 'POST' }),
    retryPipeline: (owner: string, repo: string, id: number) =>
      request<{ ok: boolean }>(`/repos/${owner}/${repo}/pipelines/${id}/retry`, { method: 'POST' }),

    // Jobs
    listJobs: (owner: string, repo: string, pipelineId: number) =>
      request<{ items: CICDJob[] }>(`/repos/${owner}/${repo}/pipelines/${pipelineId}/jobs`),
    getJob: (owner: string, repo: string, jobId: number) =>
      request<CICDJob>(`/repos/${owner}/${repo}/jobs/${jobId}`),
    getJobLogs: (owner: string, repo: string, jobId: number, fromLine = 0, limit = 5000) =>
      request<{ items: JobLog[]; fromLine: number; limit: number }>(
        `/repos/${owner}/${repo}/jobs/${jobId}/logs?from_line=${fromLine}&limit=${limit}`
      ),
    triggerManualJob: (owner: string, repo: string, jobId: number, variables?: Record<string, string>) =>
      request<{ ok: boolean }>(`/repos/${owner}/${repo}/jobs/${jobId}/trigger`, {
        method: 'POST',
        body: { variables: variables || {} },
      }),

    // Artifacts
    listArtifacts: (owner: string, repo: string, pipelineId: number) =>
      request<{ items: Artifact[] }>(`/repos/${owner}/${repo}/pipelines/${pipelineId}/artifacts`),
    // 下载 artifact:跨域请求需带 cookie,用 fetch + blob 触发浏览器下载
    downloadArtifact: async (owner: string, repo: string, artifactId: number, filename?: string) => {
      const resp = await fetch(`${API_BASE}/repos/${owner}/${repo}/artifacts/${artifactId}/download`, {
        credentials: 'include',
      })
      if (!resp.ok) throw new Error(`Download failed: ${resp.status}`)
      const blob = await resp.blob()
      const url = URL.createObjectURL(blob)
      const a = document.createElement('a')
      a.href = url
      a.download = filename || `artifact-${artifactId}`
      document.body.appendChild(a)
      a.click()
      document.body.removeChild(a)
      URL.revokeObjectURL(url)
    },

    // Secrets
    listSecrets: (owner: string, repo: string) =>
      request<{ items: Secret[] }>(`/repos/${owner}/${repo}/secrets`),
    setSecret: (owner: string, repo: string, key: string, value: string) =>
      request<{ ok: boolean }>(`/repos/${owner}/${repo}/secrets`, { method: 'POST', body: { key, value } }),
    deleteSecret: (owner: string, repo: string, key: string) =>
      request<{ ok: boolean }>(`/repos/${owner}/${repo}/secrets/${encodeURIComponent(key)}`, { method: 'DELETE' }),

    // Environments & Deployments
    listEnvironments: (owner: string, repo: string) =>
      request<{ items: Environment[] }>(`/repos/${owner}/${repo}/environments`),
    listDeployments: (owner: string, repo: string, environment?: string, page = 1, pageSize = 20) => {
      const params = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
      if (environment) params.set('environment', environment)
      return request<{ items: Deployment[]; total: number; page: number; pageSize: number }>(
        `/repos/${owner}/${repo}/deployments?${params.toString()}`
      )
    },
    listDeploymentsByCommit: (owner: string, repo: string, commitSha: string) =>
      request<{ items: Deployment[] }>(`/repos/${owner}/${repo}/commits/${commitSha}/deployments`),

    // Runners (admin)
    listRunners: () => request<{ items: Runner[] }>(`/admin/runners`),
    deleteRunner: (id: number) => request<{ ok: boolean }>(`/admin/runners/${id}`, { method: 'DELETE' }),

    // Cron Schedules (定时触发)
    listCrons: (owner: string, repo: string) =>
      request<{ items: CronSchedule[] }>(`/repos/${owner}/${repo}/crons`),
    createCron: (owner: string, repo: string, data: {
      name: string
      schedule: string
      branch: string
      yamlConfig?: string
    }) =>
      request<CronSchedule>(`/repos/${owner}/${repo}/crons`, { method: 'POST', body: data }),
    getCron: (owner: string, repo: string, id: number) =>
      request<CronSchedule>(`/repos/${owner}/${repo}/crons/${id}`),
    updateCron: (owner: string, repo: string, id: number, data: {
      schedule: string
      branch: string
      yamlConfig?: string
      enabled: boolean
    }) =>
      request<CronSchedule>(`/repos/${owner}/${repo}/crons/${id}`, { method: 'PATCH', body: data }),
    deleteCron: (owner: string, repo: string, id: number) =>
      request<void>(`/repos/${owner}/${repo}/crons/${id}`, { method: 'DELETE' }),
  }
}

File: server/internal/router/cicd.go
package router

import (
	"iforge/iforge/internal/container"
	"iforge/iforge/internal/middleware"

	"github.com/gofiber/fiber/v2"
)

// registerCICDRoutes registers CI/CD routes
//
// Route organization:
//   - Pipeline:  /repos/:owner/:repo/pipelines/*
//   - Job:       /repos/:owner/:repo/jobs/*  + /pipelines/:id/jobs
//   - Secret:    /repos/:owner/:repo/secrets/*
//   - Env:       /repos/:owner/:repo/environments
//   - Deploy:    /repos/:owner/:repo/deployments + /commits/:commit/deployments
//   - Runner:    /admin/runners (admin view)
//
// Note: External Runner API (/runners/*) is registered separately by registerCICDRunnerRoutes,
// and must be registered before the authRequired group in router.go to avoid being intercepted
// by the global RequireAuth middleware (runners use Bearer token auth, not session).
func registerCICDRoutes(api fiber.Router, c *container.Container) {
	h := c.CICDHandler()

	// Pipelines
	api.Get("/repos/:owner/:repo/pipelines", h.ListPipelines)
	api.Get("/repos/:owner/:repo/pipelines/:id", h.GetPipeline)
	api.Post("/repos/:owner/:repo/pipelines/:id/cancel", h.CancelPipeline)
	api.Post("/repos/:owner/:repo/pipelines/:id/retry", h.RetryPipeline)
	api.Get("/repos/:owner/:repo/pipelines/:id/jobs", h.ListJobs)
	api.Get("/repos/:owner/:repo/pipelines/:id/artifacts", h.ListArtifacts)

	// Jobs (independent paths for direct access by job ID)
	api.Get("/repos/:owner/:repo/jobs/:jobId", h.GetJob)
	api.Get("/repos/:owner/:repo/jobs/:jobId/logs", h.GetJobLogs)
	api.Get("/repos/:owner/:repo/jobs/:jobId/logs/stream", h.StreamJobLogs)
	api.Post("/repos/:owner/:repo/jobs/:jobId/trigger", h.TriggerManualJob)

	// Artifacts (download)
	api.Get("/repos/:owner/:repo/artifacts/:artifactId/download", h.DownloadArtifact)

	// Secrets
	api.Get("/repos/:owner/:repo/secrets", h.ListSecrets)
	api.Post("/repos/:owner/:repo/secrets", h.SetSecret)
	api.Delete("/repos/:owner/:repo/secrets/:key", h.DeleteSecret)

	// Environments & Deployments
	api.Get("/repos/:owner/:repo/environments", h.ListEnvironments)
	api.Get("/repos/:owner/:repo/deployments", h.ListDeployments)
	api.Get("/repos/:owner/:repo/commits/:commit/deployments", h.ListDeploymentsByCommit)

	// Cron Schedules (scheduled CI/CD pipeline triggers)
	api.Get("/repos/:owner/:repo/crons", h.ListCrons)
	api.Post("/repos/:owner/:repo/crons", h.CreateCron)
	api.Get("/repos/:owner/:repo/crons/:id", h.GetCron)
	api.Patch("/repos/:owner/:repo/crons/:id", h.UpdateCron)
	api.Delete("/repos/:owner/:repo/crons/:id", h.DeleteCron)

	// Runners (admin view, requires admin session)
	api.Get("/admin/runners", h.ListRunners)
	api.Delete("/admin/runners/:id", h.DeleteRunner)
}

// registerCICDRunnerRoutes registers external Runner API routes
//
// Must be called before `authRequired := api.Group("", middleware.RequireAuth())` in router.go,
// otherwise runner Bearer token requests will be intercepted by the global RequireAuth middleware (user==nil).
//
// Endpoints:
//
//	POST /runners/register            - Register runner (requires admin session)
//	POST /runners/heartbeat           - Heartbeat (Bearer token)
//	POST /runners/jobs/claim          - Claim pending job (Bearer token)
//	POST /runners/jobs/:id/status     - Report job status (Bearer token)
//	POST /runners/jobs/:id/logs       - Batch upload logs (Bearer token)
func registerCICDRunnerRoutes(api fiber.Router, c *container.Container) {
	rh := c.CICDRunnerHandler()
	runnerAuth := middleware.RunnerAuthMiddleware(c.RunnerService())

	// register requires admin session (AuthMiddleware parses session + RequireAdmin enforces admin)
	api.Post("/runners/register", middleware.RequireAdmin(), rh.RegisterRunner)

	// The following endpoints use runner Bearer token authentication (not session)
	api.Post("/runners/heartbeat", runnerAuth, rh.Heartbeat)
	api.Post("/runners/jobs/claim", runnerAuth, rh.ClaimJob)
	api.Post("/runners/jobs/:id/status", runnerAuth, rh.UpdateJobStatus)
	api.Post("/runners/jobs/:id/logs", runnerAuth, rh.AppendJobLogs)
}
